src/main/scala/server/ServerApp.scala

```diff
@@ -8,40 +8,61 @@ import org.http4s.dsl.io._
 import org.http4s.implicits._
 import io.circe.generic.auto._
 import org.http4s.circe._
-import Domain.{Subscription, SubscriptionRequest}
+import Domain.{Subscription, SubscriptionRequest, PutRequest}
 
 import java.util.UUID
 import scala.concurrent.ExecutionContext.global
 
 object ServerApp extends IOApp {
 
-  implicit val decoder: EntityDecoder[IO, SubscriptionRequest] = jsonOf[IO, SubscriptionRequest]
+  implicit val decoderPost: EntityDecoder[IO, SubscriptionRequest] = jsonOf[IO, SubscriptionRequest]
+  implicit val decoderPut: EntityDecoder[IO, PutRequest] = jsonOf[IO, PutRequest]
 
   override def run(args: List[String]): IO[ExitCode] = {
-    var storage: Map[String, Subscription] = Map.empty
+    implicit var storage: Map[String, Subscription] = Map.empty
+
+    def existsId(id: String, request: Either[String, PutRequest])(implicit storage: Map[String, Subscription]): Either[String, PutRequest] = {
+      if (!storage.contains(id))
+        Left(s"The subscription with id $id does not exist + ${request.left.toOption.get}")
+      else
+        request
+    }
 
     val helloWorldService: Kleisli[IO, Request[IO], Response[IO]] = HttpRoutes.of[IO] {
       case GET -> Root / "hello" / name =>
         Ok(s"Hello, $name.")
       case DELETE -> Root / "hello" / name =>
         Ok(s"No hello for $name")
 
-      case req@POST -> Root / "subscriptions" => for {
+      case req @ POST -> Root / "subscriptions" => for {
         request <- req.as[SubscriptionRequest]
         maybeValidRequest = SubscriptionRequest.validate(request)
         _ <- IO(println(maybeValidRequest))
-        _ <- IO(println(storage))
         result <- maybeValidRequest match {
           case None => BadRequest("Invalid body")
           case Some(request) =>
             val id = UUID.randomUUID().toString
             val subscription = Subscription(id, request.origin, request.destination, request.label)
             storage = Map(id -> subscription) ++ storage
+            println(storage)
             Ok(s"""{"id": $id}""")
         }
       } yield result
 
-      case req @ PUT -> Root / "subscriptions" / id => ???
+      case req @ PUT -> Root / "subscriptions" / id => for {
+        request <- req.as[PutRequest]
+        maybeValidRequest = PutRequest.validate(request)
+        _ <- IO(println(maybeValidRequest))
+        request = existsId(id, maybeValidRequest)
+        result <- request match {
+          case Left(value) => BadRequest(value)
+          case Right(request) =>
+            val subscription = Subscription(id, request.origin, request.destination, request.label)
+            storage = storage ++ Map(id -> subscription)
```

Comment:
I like how you come up with idea to update the subscription in the `Map` 👍

Reply:
yeah, I realized that order is important 😄